kube-core/src/metadata.rs

```diff
@@ -12,3 +13,124 @@ pub struct TypeMeta {
     /// The name of the API
     pub kind: String,
 }
+
+/// A runtime accessor trait for TypeMeta
+///
+/// This trait can be thought of as a limited variant of the `Resource` trait that reads from runtime properties.
+/// It cannot retrieve the plural, nor the scope of a resource and requires an `ApiResource` for this instead.
+///
+/// For static types is generally leans on the static information, but for dynamic types, it inspects the object.
+pub trait TypeInfo {
+    /// Get the `TypeMeta` of an object
+    ///
+    /// This is a safe `TypeMeta` getter for all object types
+    fn types(&self) -> Option<TypeMeta>;
+
+    /// Get the `TypeMeta` of an object that is guaranteed to have it
+    ///
+    /// Returns `TypeMeta` when exists, panics otherwise
+    fn types_unchecked(&self) -> TypeMeta;
+
+    /// Get the `kind` of an object
+    fn kind(&self) -> Option<String>;
+    /// Get the `apiVersion` of any object
+    fn api_version(&self) -> Option<String>;
```

Comment:
Would it make sense to return either `&str` or `Cow<str>` rather than `String` here? People who want to take ownership can still do so trivially.

Reply:
Happy to do that 👍

Reply:
Fixed in https://github.com/kube-rs/kube/pull/1053/commits/05b85404299d590a4d282b9f6b8495b71ee85732